File: python/Problem39.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-
'''
If p is the perimeter of a right angle triangle with integral length sides, {a,b,c}, there are exactly three solutions for p = 120.

{20,48,52}, {24,45,51}, {30,40,50}

For which value of p  1000, is the number of solutions maximised?
'''
from math import hypot

def solution():
    for i in range(12, 1000, 12):
        acc = []
        for s in range(1, i):
            for t in range(s, i - s):
                u = hypot(s, t)
                if u.is_integer() and s + t + u == i:
                    acc.append((s, t, int(u)))
        if acc:
            yield len(acc), i


if __name__ == '__main__':
    print max(solution())[1]

File: scala/Problem39.scala
/*
If p is the perimeter of a right angle triangle with integral length sides, {a,
b,c}, there are exactly three solutions for p = 120.

{20,48,52}, {24,45,51}, {30,40,50}

For which value of p ≤ 1000, is the number of solutions maximised?
*/

object Problem39 {
  def findPytagorean(perimeter:Int) = {
    (1 to perimeter-1).flatMap(a => (1 to a).map(b => List(b, a, math.sqrt(a*a+b*b))).filter(_.sum == perimeter))
  }
  def main(args:Array[String]) {
    println((1 to 1000).map(n => (findPytagorean(n).length, n)).max._2)
  }
}

